### ISSUE TITLE
LFS Track Files (and other operations) as Go Library

### ISSUE THREAD

**Issue by Southclaws** [git-lfs/git-lfs#4040]
**Describe the issue**

I want to perform certain actions such as `git lfs track` from Go code, not executing a command. I've noticed that the behaviour of `track` is rather tightly coupled to the argument parsing/command interface. (https://github.com/git-lfs/git-lfs/blob/master/commands/command_track.go#L118)

**System environment**

I'm interested in making a little tool that's cross platform and self-contained - so no dependency on Git.exe, Golang is perfect for this role and it would be great if I could ship _one_ binary instead of shipping mine _and_ `git-lfs.exe` and doing `cmd.Exec`.

---

I'd be interested in contributing some changes to make this more library-friendly if that's something that would be accepted in a PR.

If not, that's fine, I can just build two executables and call one with `cmd.Exec` 👍

Thanks!

**Comment by bk2204** [comment]
Hey,

Speaking personally, I'm not opposed to seeing pieces split out into logical packages if that's more convenient for folks.  However, you should be aware that we wouldn't consider those packages API stable or immune from gaining dependencies, so you might find that they changed over time in an inconvenient way.  We also do make heavy use of the Git binary in places, so that may be problematic for you.  I have an Go project for $DAYJOB that is explicitly not using parts of Git LFS for those reasons.

However, if you're okay with those restrictions, I'm happy to review PRs.

**Comment by J0Nes90** [comment]
Hey,

any update on this ticket @Southclaws ? Are you still working on this?
I would love to see such library as well. I am using libgit2 extensively and for that purpose it is far from ideal that git-lfs invokes git itself. 

I was able to compile git-lfs as a C library and to use it from within my application. I wonder if it would make sense to replace the git commands inside the Git LFS with an interface that abstracts calls to libgit2 or git, for example. But I can understand that you want to keep the dependencies minimal. Maybe it would be possible to set a custom path to a git installation (so the git application could be part of a bundle, e.g. on macOS)

**Comment by Southclaws** [comment]
No, I don't have time, I just call the command line utility or use Git. I do plan to implement proper clean/smudge filters into [go-git](https://github.com/go-git/go-git/) eventually though, which makes integration a little easier.
